src/node.rs

```diff
@@ -270,6 +273,9 @@ pub struct InMemoryNodeInner<S> {
     pub console_log_handler: ConsoleLogHandler,
     pub system_contracts: SystemContracts,
     pub impersonated_accounts: HashSet<Address>,
+    /// Keeps track of historical states indexed via block hash.
+    /// This is currently unbounded and may change in future to keep only the last N blocks.
+    pub previous_states: HashMap<H256, HashMap<StorageKey, StorageValue>>,
```

Comment:
Unsure if we should already do it or not care about it at all?

Reply:
I'm guessing you're referring to the bounding to N blocks? Yeah, this one does slightly worry me for size as we'll be cloning the raw_storage every time. If someone keeps this node up for too long, this HashMap can grow exponentially.

I'm fine letting this go as-is, and just leave this comment to watch out for future space/performance issues.

Reply:
in foundry they keep a .json cache of StorageMap per block- maybe it's worth copying this pattern and using a LRU to keep some bounded number of blocks in memory?

Reply:
The question would be to how long (if at all) to persist it given that it's a test node. We could remove all previous state on start with a disk backed LRU. Just wondering if it's needed and if yes, to what degree since ideally the state shouldn't be huge. Perhaps a follow up ticket to revisit this.

Reply:
Agreed that if we persist all state on disk, then we'd probably need to remove all files on restart of the node. Perhaps we only cache storage requests to upstream for forks?

I noticed that we're not cache-ing within the `get_storage_at` function of `HttpForkSource`

Reply:
Yeah that's cause initially I decided it would be hard to cache against `latest` block. I think it might be even hard to resolve the latest block number at the `HttpForkSource` given it doesn't know of the node.

Reply:
Hmmm yeah :/ ... Kk, it seems like the path forward for now is to leave as is and revisit if problems arise. For what it's worth, I pulled down this branch and ran the e2e-tests dozens of times with no impact in response time from the node.  I'm sure Memory usage went up, but I'm not overly worried about that at the moment.

We can also discuss offline tomorrow if that helps

Reply:
PR updated with 128 block FIFO queue of storage